File: VC_HIL/python/inverterhilgui/heartbeat_state.py
import math
from .sequence_command import sequenceCommand


def heartbeatState(previous, nowS, timeoutS):
    result = {"counter": 0, "lastUpdateS": math.nan, "ageS": math.inf, "expired": True, "reason": "malformed_previous_state"}
    if not isinstance(previous, dict) or "counter" not in previous or "lastUpdateS" not in previous: return result
    counter = previous["counter"]
    if not isinstance(counter, (int, float, bool)) or not math.isfinite(float(counter)) or counter < 0 or counter > 4294967295 or float(counter) != int(counter):
        result["reason"] = "malformed_counter"; return result
    result["counter"] = int(counter)
    if not isinstance(nowS, (int, float)) or isinstance(nowS, bool) or not math.isfinite(nowS): result["reason"] = "malformed_now"; return result
    if not isinstance(timeoutS, (int, float)) or isinstance(timeoutS, bool) or not math.isfinite(timeoutS) or timeoutS <= 0: result["reason"] = "malformed_timeout"; return result
    last = previous["lastUpdateS"]
    if not isinstance(last, (int, float)) or isinstance(last, bool) or not math.isfinite(last):
        result.update(counter=sequenceCommand(counter), lastUpdateS=float(nowS), reason="no_previous_beat"); return result
    age = float(nowS) - float(last)
    if age < 0: result["reason"] = "host_clock_went_backwards"; return result
    expired = age > timeoutS
    result.update(counter=sequenceCommand(counter), lastUpdateS=float(nowS), ageS=age, expired=expired,
                  reason="heartbeat_expired" if expired else "healthy")
    return result

File: VC_HIL/inverter/+inverterhilgui/+writes/heartbeatState.m
function result = heartbeatState(previous, nowS, timeoutS)
%HEARTBEATSTATE Advance the GUI heartbeat counter and evaluate its age.
%
%   RESULT = HEARTBEATSTATE(PREVIOUS, NOWS, TIMEOUTS) is called once per app
%   timer tick. PREVIOUS is a scalar structure with fields COUNTER (uint32)
%   and LASTUPDATES (host seconds). RESULT.COUNTER is the wrap-safe uint32
%   value to write to HIL_CMD.GUI_HEARTBEAT and RESULT.AGES is the interval
%   since the previous beat, so a stalled host timer is visible to the
%   operator as well as to the target.
%
%   This evaluator FAILS CLOSED: malformed input or a missing previous state
%   reports EXPIRED true and holds the counter at its last known good value,
%   because an unknown heartbeat must be treated as a lost heartbeat.

result = struct( ...
    'counter', uint32(0), ...
    'lastUpdateS', NaN, ...
    'ageS', Inf, ...
    'expired', true, ...
    'reason', 'malformed_previous_state');

if ~isstruct(previous) || ~isscalar(previous) || ...
        ~isfield(previous, 'counter') || ~isfield(previous, 'lastUpdateS')
    return;
end
counterValue = previous.counter;
if ~(isnumeric(counterValue) || islogical(counterValue)) || ...
        ~isscalar(counterValue) || ~isreal(counterValue) || ...
        ~isfinite(double(counterValue)) || double(counterValue) < 0 || ...
        double(counterValue) > 4294967295 || ...
        double(counterValue) ~= floor(double(counterValue))
    result.reason = 'malformed_counter';
    return;
end
result.counter = uint32(counterValue);

if ~isnumeric(nowS) || ~isscalar(nowS) || ~isreal(nowS) || ~isfinite(nowS)
    result.reason = 'malformed_now';
    return;
end
if ~isnumeric(timeoutS) || ~isscalar(timeoutS) || ~isreal(timeoutS) || ...
        ~isfinite(timeoutS) || timeoutS <= 0
    result.reason = 'malformed_timeout';
    return;
end

lastUpdateS = previous.lastUpdateS;
if ~isnumeric(lastUpdateS) || ~isscalar(lastUpdateS) || ...
        ~isreal(lastUpdateS) || ~isfinite(lastUpdateS)
    result.counter = inverterhilgui.writes.sequenceCommand(result.counter);
    result.lastUpdateS = double(nowS);
    result.ageS = Inf;
    result.expired = true;
    result.reason = 'no_previous_beat';
    return;
end

ageS = double(nowS) - double(lastUpdateS);
if ageS < 0
    result.ageS = Inf;
    result.expired = true;
    result.reason = 'host_clock_went_backwards';
    return;
end

result.counter = inverterhilgui.writes.sequenceCommand(result.counter);
result.lastUpdateS = double(nowS);
result.ageS = ageS;
result.expired = ageS > double(timeoutS);
if result.expired
    result.reason = 'heartbeat_expired';
else
    result.reason = 'healthy';
end
end
